backend/bridgecommon/bridgecommon.go

```diff
@@ -93,12 +96,23 @@ func BackendCall(queryID int, jsonQuery string) {
 			panic(errp.WithStack(err))
 		}
 		request.Header.Set("Authorization", "Basic "+token)
-		handlers.Router.ServeHTTP(resp, request)
+		handlersInstance.Router.ServeHTTP(resp, request)
 		responseBytes := resp.Body.Bytes()
 		communication.Respond(queryID, string(responseBytes))
 	}()
 }
 
+// UsingMobileDataChanged should be called when the network connnection changed.
+func UsingMobileDataChanged() {
+	if backendInstance == nil {
```

Comment:
Isn't this a concurrent access issue? Assuming it can be read here and set in `Serve` func below on line 173 at the same time.

If it isn't possible, would remove this nil check and assume it cannot be nil at the time `UsingMobileDataChanged` is called.

Reply:
In Java world I don't check that this function is not called before Serve() is called. Same deal for BackendCall(). Looked easier to check it here instead. WDYT?

Reply:
Yeah, I think it would look ugly to trace what's called first in Java at runtime.

My problem is with the `backendInstance == nil` check existence here. The check tells me it's possible for it to be `nil` but I don't think it's the case and so it confuses me. If it were possible though, it would be a race condition between here and in `Serve` func below on line 173 where it's set with `backendInstance, err = backend.NewBackend(...)`, in which case `backendInstance` should've been guarded with a mutex.

In the end, I don't think it is possible for `backendInstance` to be nil and so there's no need for `if backendInstance == nil` at all. The `backendInstance` is created in `Serve` which is called from `GoViewModel` in its constructor https://github.com/digitalbitbox/bitbox-wallet-app/blob/a860d64f/frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/GoViewModel.java#L185. And the `GoViewModel` is created only once at the app startup or so it says in https://github.com/digitalbitbox/bitbox-wallet-app/blob/fd0c7fcd/frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java#L102-L104

Anyway, my point is, `backendInstance` should either be guarded with a mutex or there's no need for `nil` check at all because it can't be. I believe it's the latter, not the former!

Looks the same for BackendCall to me but we can deal with it later.

Reply:
The function is called from the network broadcast receiver. It sounds hard to reason about it, but to me it sounds like that could happen at any time, even before Serve() is called. Maybe I am wrong. In any case, this is what I meant: I'd rather have the check here than try to reason about the Android SDK or our Java code.

I can add a mutex, though I don't understand how it helps? If this function is called before Serve() (race condition or not), it just doesn't do anything, but that isn't bad.

I'll add one anyway, I guess sometimes a mutex is just expected to not have to think about it.

Reply:
Well, when a race condition is possible, a mutex prevents that which otherwise can cause a panic and crash the app. Here's a simple example. I put the output at the end: https://play.golang.org/p/ojnIQVN9GYZ

Reply:
> The function is called from the network broadcast receiver. It sounds hard to reason about it, but to me it sounds like that could happen at any time, even before Serve()

The receiver is indeed created in the app constructor but it is registered to receive events only in `onResume` in https://github.com/digitalbitbox/bitbox-wallet-app/blob/fd0c7fcd/frontends/android/BitBoxApp/app/src/main/java/ch/shiftcrypto/bitboxapp/MainActivity.java#L247 which is [always called after](https://developer.android.com/guide/components/activities/activity-lifecycle) `onCreate`, the place where `backendInstance` is eventually created.

Reply:
OTOH if `backendInstance` access is moved elsewhere and used concurrently in some refactoring, we will most likely forget to put a mutex on it. So, I would add it now to be future safe.

Reply:
Makes sense of course, not sure what I was thinking ;)

Added a commit.